### PR TITLE
feat: support flat config and eslint v9

### PR DESCRIPTION
**Pull request by kazupon**
## background

The flat config will be supported officially in eslint v9 release, and the current config style will be removed in eslint v10.
We can already use flat config in recent versions of eslint v8.

https://eslint.org/docs/latest/use/configure/configuration-files-new

And some deprecated APIs was dropped :

https://eslint.org/blog/2023/09/preparing-custom-rules-eslint-v9/

## about implementation

This PR supports the existing config style while also supporting flat config, and eslint v9 support

## PS
I've just supported eslint v9 and eslint v8.x compatibility: 2024-04-11

### CODE REVIEW

File: docs/.vitepress/config.mts
Comment (kazupon) [inline docs/.vitepress/config.mts:1]: avoid vitepress (vite) trouble 
https://vitejs.dev/guide/troubleshooting.html#this-package-is-esm-only

File: src/configs/flat/recommended.ts
Comment (kazupon) [inline src/configs/flat/recommended.ts:1]: I've referenced eslint-plugin-vue for implementation.
https://github.com/vuejs/eslint-plugin-vue/pull/2407
https://github.com/vuejs/eslint-plugin-vue/pull/2436

File: tsconfig.json [hunk 1]
```diff
@@ -6,6 +6,7 @@
     "strict": true,
     "esModuleInterop": true,
     "resolveJsonModule": true,
+    "declaration": true,
```
Comment (kazupon) [inline tsconfig.json:9]: To support DX in flat config, types are now also generated.

**Comment by kazupon** [comment]
I've just supported eslint v9 and eslint v8.x compatibility

**Comment by vhoyer** [comment]
release on `2.3.0`

**Comment by kazupon** [comment]
Thanks!

**Pull request merged**